Enable testing on Windows

elopez opened the pull request:
## Description

This (mostly) fixes hevm (or should I say libff 😅 ) on Windows and enables CI testing.

## Checklist

- [ ] tested locally
- [ ] added automated tests
- [ ] updated the docs
- [ ] updated the changelog

d-xo commented on test/rpc.hs:
```diff
@@ -69,7 +75,7 @@ tests = testGroup "rpc"
   , testGroup "execution with remote state"
     -- execute against remote state from a ds-test harness
-    [ test "dapp-test" $ do
+    [ ignoreTestWindows $ test "dapp-test" $ do
```
why don't these work on windows?

elopez replied:
Some of the git invocations that hevm did were failing and it was not obvious to me why. There's a comment with the error in the commit that added this ignore.

d-xo replied:
could we just drop a source level comment here to make the reasoning here a little more accesible?

d-xo commented on test/test.hs:
curious into why the tests that are ignored on windows here are failing there?

elopez replied:
I think they all use bitwuzla, which I think is not supported on Windows. If I'm mistaken and it is and there's binaries somewhere, we can install it in CI and run these though.

d-xo replied:
ah, perhaps we can just conditionally use cvc5 or z3 on windows for these tests or something?

d-xo commented on .github/scripts/libff.patch:
curious as to the contents of the patch here?

elopez replied:
The gist of the patch is that libff uses mpz_*_ui functions (that return unsigned long) to operate on 64 bits of data at a time. This works on Linux/Max because unsigned long is indeed 64 bits there. But on Windows, unsigned long is 32 bits, so the functions don't work as expected. This patch rewrites them to behave correctly, by mainly using functions that operate on "limbs" which are the right size.

msooseth commented on .github/scripts/libff.patch:
```diff
@@ -0,0 +1,215 @@
+-            return ((i+1) * GMP_NUMB_BITS) - __builtin_clzl(x);
++            if(sizeof(x) == sizeof(unsigned long))
++                return ((i+1) * GMP_NUMB_BITS) - __builtin_clzl(x);
++            else
```
I'd probably put an `assert` here to make sure it's `sizeof(unsigned long long)`. But maybe I'm too cautious.

msooseth commented:
Wow, this is some serious work, thanks! Is this ready or is there a specific reason why it's marked for `work in progress`? Is there something I can help with?

elopez commented:
@msooseth I assumed you may not want to carry a patch on libff in the repo so I left it marked as WIP, but if you're ok with it I guess we can consider this ready for review 👍

msooseth commented:
I mean, upstreaming it to libff would probably be best :) But we can have this in the repo as long as it works I guess? Due to nix we don't automatically update libff unless really necessary.

elopez commented:
I resolved the conflicts and refreshed the branch after the 9.6 merge 😄  I also added some more workflow code to build bitwuzla for Windows, so now those tests also run on Windows CI 🧪 The built binaries are cached so as to not slow down the CI too much.

Test-wise it'd be nice to see why the dapp/foundry Git stuff fails on Windows, then Windows would be at test parity with the other platforms. In the meantime I added a reason string to the ignore function.

I also had to add a new flag to the cabal file, as the newer secp256k1 now requires a new #define to be set so as to link correctly when building hevm against the static library on Windows.

msooseth commented:
Thanks! Looks nice. If it ever breaks, you may need to help, as none of us are Windows experts. But I hope it won't break too often :) Thank you again,

Mate
